lib/MonadicRewrite.thy

```diff
@@ -14,24 +15,100 @@ imports
   LemmaBucket
 begin
 
-definition
-  monadic_rewrite :: "bool \<Rightarrow> bool \<Rightarrow> ('a \<Rightarrow> bool)
-                       \<Rightarrow> ('a, 'b) nondet_monad \<Rightarrow> ('a, 'b) nondet_monad \<Rightarrow> bool"
-where
- "monadic_rewrite F E P f g \<equiv> \<forall>s. P s \<and> (F \<longrightarrow> \<not> snd (f s))
-           \<longrightarrow> (E \<longrightarrow> f s = g s)
-            \<and> (\<not> E \<longrightarrow> fst (g s) \<subseteq> fst (f s) \<and> (snd (g s) \<longrightarrow> snd (f s)))"
+definition monadic_rewrite ::
+  "bool \<Rightarrow> bool \<Rightarrow> ('s \<Rightarrow> bool) \<Rightarrow> ('s, 'a) nondet_monad \<Rightarrow> ('s, 'a) nondet_monad \<Rightarrow> bool"
+  where
+   "monadic_rewrite F E P f g \<equiv>
+      \<forall>s. P s \<and> (F \<longrightarrow> \<not> snd (f s))
+          \<longrightarrow> (E \<longrightarrow> f s = g s) \<and> (\<not> E \<longrightarrow> fst (g s) \<subseteq> fst (f s) \<and> (snd (g s) \<longrightarrow> snd (f s)))"
 
+lemma monadic_rewrite_inst:
+  "monadic_rewrite F E P f g \<Longrightarrow> monadic_rewrite F E P f g"
+  by assumption
+
+lemma monadic_rewrite_refl:
+  "monadic_rewrite F E \<top> f f"
```

Comment:
Also an idea for later: instead of \top I'd have expected any P here. Of course with weakening this is equivalent, and maybe we actually want P to be as concrete as possible for "final" rules in terms of rewriting. But in terms of basic properties, I'd do a `P` version and potentially use an instance with \top for rewriting.

Reply:
I'll add it, but will make it `monadic_rewrite_refl_gen` for now (better name welcome) because `monadic_rewrite_refl` is used a lot in its `\top` form and I don't want that rename turbulence on this PR. I suspect it's more useful in the `\top` form anyway.

Reply:
Done